packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+import * as asyncHooks from 'async_hooks';
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = Object.create(null);
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      destroy: this.destroy.bind(this),
+      promiseResolve: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.scopes[asyncHooks.executionAsyncId()] || null;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: T
+  ): ReturnType<T> {
+    const uid = asyncHooks.executionAsyncId();
+    const oldScope = this.scopes[uid] || null;
+    this.scopes[uid] = scope;
+    try {
+      const res = fn();
+      this.scopes[uid] = oldScope;
+      return res;
+    } catch (err) {
+      // restore old scope even if the function throw
+      this.scopes[uid] = oldScope;
+      throw err;
+    }
+  }
+
+  bind<T>(target: T, scope?: unknown): T {
+    // if no specific scope to propagate is given, we use the current one
+    if (!scope) {
```

Comment:
This should be an explicit check for `undefined` to allow binding to `null` explicitly. This can happen if `bind` is called while there is no scope currently.